server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java

```diff
@@ -23,15 +26,24 @@
  */
 public class BroadcastUnpromotableRequest extends ActionRequest {
 
+    /**
+     * Holds the index shard routing table that will be used by {@link TransportBroadcastUnpromotableAction} to broadcast the requests to
+     * the unpromotable replicas. The routing table is not serialized over the wire, and will be null on the other end of the wire.
+     * For this reason, the field is package-private.
+     */
+    final @Nullable IndexShardRoutingTable indexShardRoutingTable;
```

Comment:
Should it be private?

Reply:
Hi @idegtiarenko , no, because it's used `TransportBroadcastUnpromotableAction#doExecute` and that is why I wanted it to be package-private (as mentioned in the comment). Feel free to review the comment.

Reply:
There's not much more to an `IndexShardRoutingTable` than just the `ShardRouting` things it contains. I'd rather keep the whole ISRT here - if we don't have the `ShardRouting` objects then we can't filter out the promotable shards within the action, we have to ask every caller to do that, and the whole point of this PR is to centralise that logic.

Reply:
I also believe we may be more flexible with the whole IndexShardRoutingTable in the future if ever needed.

Reply:
@tlrx You mean that the cluster state might already be updated on the local primary node during the replication action, and we can now detect it here. But what would happen if it's not of a higher version? Would it throw or proceed somehow with the latest cluster state? In this case, it proceeds with the index shard routing table of the older version. If it would proceed with the most recent cluster state, then we'd end up not using the ISRT that @DaveCTurner wanted. If it would throw, then that's a new behavior that I can implement, but I wonder if we want that? It would mean that the promotable replicas might proceed (with the older replication group) while the unpromotables would fail because we're not on the latest cluster state. I find this discrepancy a bit weird.

Reply:
From @tlrx :

> I've been puzzled by the fact that we want to reuse the unpromotable shards from the same replication group as the original replication action (and hence we must locally pass the ISRT around)... but a responsability of replication action is to mark shards as stale, ie removed them from the in sync set, and unpromotable shards do not belong to the in-sync set and as such why does this matter to use the same ISRT => hence my question if a "fresh" enough ISRT could be enough. But I got the point of to keep it close of the TransportReplicationAction, I see how it can help for future action. I'm not going to slow you down more, thanks both for answering.

@DaveCTurner I also saw your answer.

I think from your comments you both do not require anything additional in the PR at the moment in this aspect. But feel free to tell me if I misunderstood and you prefer the way of somehow checking the cluster state version.

Thanks!